File: backend/internal/repository/employee_central.go
package repository

import (
	"context"
	"time"

	"hr-saas/internal/db"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type EmployeeDocument struct {
	ID           uuid.UUID  `json:"id"`
	TenantID     uuid.UUID  `json:"tenant_id"`
	EmployeeID   uuid.UUID  `json:"employee_id"`
	DocumentName string     `json:"document_name"`
	DocumentType string     `json:"document_type"`
	DocumentURL  string     `json:"document_url"`
	UploadDate   time.Time  `json:"upload_date"`
	ExpiryDate   *time.Time `json:"expiry_date"`
	Status       string     `json:"status"`
	CreatedAt    time.Time  `json:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at"`
	// Join
	EmployeeName string `json:"employee_name,omitempty"`
}

type ProbationRecord struct {
	ID         uuid.UUID  `json:"id"`
	TenantID   uuid.UUID  `json:"tenant_id"`
	EmployeeID uuid.UUID  `json:"employee_id"`
	StartDate  time.Time  `json:"start_date"`
	EndDate    time.Time  `json:"end_date"`
	Status     string     `json:"status"`
	ReviewDate *time.Time `json:"review_date"`
	Notes      *string    `json:"notes"`
	CreatedAt  time.Time  `json:"created_at"`
	UpdatedAt  time.Time  `json:"updated_at"`
	// Join
	EmployeeName string `json:"employee_name,omitempty"`
}

type EmployeePromotion struct {
	ID                  uuid.UUID `json:"id"`
	TenantID            uuid.UUID `json:"tenant_id"`
	EmployeeID          uuid.UUID `json:"employee_id"`
	PreviousDesignation string    `json:"previous_designation"`
	NewDesignation      string    `json:"new_designation"`
	PreviousSalary      float64   `json:"previous_salary"`
	NewSalary           float64   `json:"new_salary"`
	Type                string    `json:"type"`
	EffectiveDate       time.Time `json:"effective_date"`
	Notes               *string   `json:"notes"`
	Status              string    `json:"status"`
	CreatedAt           time.Time `json:"created_at"`
	UpdatedAt           time.Time `json:"updated_at"`
	// Join
	EmployeeName string `json:"employee_name,omitempty"`
}

// ---------------------------------------------------------
// DOCUMENTS
// ---------------------------------------------------------

func ListEmployeeDocuments(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]EmployeeDocument, error) {
	var list []EmployeeDocument
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT d.id, d.tenant_id, d.employee_id, d.document_name, d.document_type, d.document_url, d.upload_date, d.expiry_date, d.status, d.created_at, d.updated_at,
			        e.name as employee_name
			 FROM employee_documents d
			 JOIN employees e ON d.employee_id = e.id
			 ORDER BY d.upload_date DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r EmployeeDocument
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.DocumentName, &r.DocumentType, &r.DocumentURL, &r.UploadDate, &r.ExpiryDate, &r.Status, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

// ---------------------------------------------------------
// PROBATION
// ---------------------------------------------------------

func ListProbationTracking(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]ProbationRecord, error) {
	var list []ProbationRecord
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT p.id, p.tenant_id, p.employee_id, p.start_date, p.end_date, p.status, p.review_date, p.notes, p.created_at, p.updated_at,
			        e.name as employee_name
			 FROM probation_tracking p
			 JOIN employees e ON p.employee_id = e.id
			 ORDER BY p.end_date ASC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r ProbationRecord
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.StartDate, &r.EndDate, &r.Status, &r.ReviewDate, &r.Notes, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

// ---------------------------------------------------------
// PROMOTIONS
// ---------------------------------------------------------

func ListEmployeePromotions(ctx context.Context, pool *pgxpool.Pool, tenantID string) ([]EmployeePromotion, error) {
	var list []EmployeePromotion
	err := db.WithTenant(ctx, pool, tenantID, func(ctx context.Context, tx pgx.Tx) error {
		rows, err := tx.Query(ctx,
			`SELECT p.id, p.tenant_id, p.employee_id, p.previous_designation, p.new_designation, p.previous_salary, p.new_salary, p.type, p.effective_date, p.notes, p.status, p.created_at, p.updated_at,
			        e.name as employee_name
			 FROM employee_promotions p
			 JOIN employees e ON p.employee_id = e.id
			 ORDER BY p.effective_date DESC`)
		if err != nil {
			return err
		}
		defer rows.Close()
		for rows.Next() {
			var r EmployeePromotion
			if err := rows.Scan(&r.ID, &r.TenantID, &r.EmployeeID, &r.PreviousDesignation, &r.NewDesignation, &r.PreviousSalary, &r.NewSalary, &r.Type, &r.EffectiveDate, &r.Notes, &r.Status, &r.CreatedAt, &r.UpdatedAt, &r.EmployeeName); err != nil {
				return err
			}
			list = append(list, r)
		}
		return rows.Err()
	})
	return list, err
}

File: frontend/lib/data/models/employee_central.dart
class EmployeeDocument {
  final String id;
  final String tenantId;
  final String employeeId;
  final String documentName;
  final String documentType;
  final String documentUrl;
  final DateTime uploadDate;
  final DateTime? expiryDate;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Joined field
  final String? employeeName;

  EmployeeDocument({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.documentName,
    required this.documentType,
    required this.documentUrl,
    required this.uploadDate,
    this.expiryDate,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.employeeName,
  });

  factory EmployeeDocument.fromJson(Map<String, dynamic> json) {
    return EmployeeDocument(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      documentName: json['document_name'] as String,
      documentType: json['document_type'] as String? ?? 'Other',
      documentUrl: json['document_url'] as String,
      uploadDate: DateTime.parse(json['upload_date'] as String),
      expiryDate: json['expiry_date'] != null ? DateTime.tryParse(json['expiry_date'] as String) : null,
      status: json['status'] as String? ?? 'active',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      employeeName: json['employee_name'] as String?,
    );
  }
}

class ProbationRecord {
  final String id;
  final String tenantId;
  final String employeeId;
  final DateTime startDate;
  final DateTime endDate;
  final String status;
  final DateTime? reviewDate;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Joined field
  final String? employeeName;

  ProbationRecord({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.startDate,
    required this.endDate,
    required this.status,
    this.reviewDate,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    this.employeeName,
  });

  factory ProbationRecord.fromJson(Map<String, dynamic> json) {
    return ProbationRecord(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      startDate: DateTime.parse(json['start_date'] as String),
      endDate: DateTime.parse(json['end_date'] as String),
      status: json['status'] as String? ?? 'active',
      reviewDate: json['review_date'] != null ? DateTime.tryParse(json['review_date'] as String) : null,
      notes: json['notes'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      employeeName: json['employee_name'] as String?,
    );
  }
}

class EmployeePromotion {
  final String id;
  final String tenantId;
  final String employeeId;
  final String previousDesignation;
  final String newDesignation;
  final double previousSalary;
  final double newSalary;
  final String type;
  final DateTime effectiveDate;
  final String? notes;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  // Joined field
  final String? employeeName;

  EmployeePromotion({
    required this.id,
    required this.tenantId,
    required this.employeeId,
    required this.previousDesignation,
    required this.newDesignation,
    required this.previousSalary,
    required this.newSalary,
    required this.type,
    required this.effectiveDate,
    this.notes,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.employeeName,
  });

  factory EmployeePromotion.fromJson(Map<String, dynamic> json) {
    return EmployeePromotion(
      id: json['id'] as String,
      tenantId: json['tenant_id'] as String,
      employeeId: json['employee_id'] as String,
      previousDesignation: json['previous_designation'] as String? ?? '',
      newDesignation: json['new_designation'] as String? ?? '',
      previousSalary: (json['previous_salary'] as num?)?.toDouble() ?? 0.0,
      newSalary: (json['new_salary'] as num?)?.toDouble() ?? 0.0,
      type: json['type'] as String? ?? 'Promotion',
      effectiveDate: DateTime.parse(json['effective_date'] as String),
      notes: json['notes'] as String?,
      status: json['status'] as String? ?? 'pending',
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      employeeName: json['employee_name'] as String?,
    );
  }
}
